hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFileSystem.java

```diff
@@ -1204,6 +1204,7 @@ public FileStatus getFileStatus(Path f) throws IOException {
     public FileStatus[] listStatus(Path f) throws AccessControlException,
```

Comment:
Overall looks good to me now.
However, This will be a behavioral change. Users may suddenly see additional dirs listed in ls result. So, I would suggest to update Javadoc for this API? ( do the same thing at ViewFs.java as well)
Ex:
```
/**
   * {@inheritDoc}
   * 
   * Note: ls on root("/") considers listing from fallbackLink also if
   * available. Also write about shading part when same dirs in mount link and
   * fallback dir?
   */
  @Override
  public FileStatus[] listStatus(final Path f) throws AccessControlException, 
```
I will also mark incompatible change at JIRA and can update release notes. What do you say?
Thanks for handling all the feedback. After adding this, I think we should be good to go.